packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { VM, scheduleRehydration } from './vm';
+import { isFalse, isUndefined, isNull } from '@lwc/shared';
+import { markComponentAsDirty, ComponentConstructor } from './component';
+import { Template } from './template';
+import { StylesheetFactory } from './stylesheet';
+
+const swappedTemplateMap = new WeakMap<Template, Template>();
+const swappedComponentMap = new WeakMap<ComponentConstructor, ComponentConstructor>();
+const swappedStyleMap = new WeakMap<StylesheetFactory, StylesheetFactory>();
+
+const activeTemplates = new WeakMap<Template, Set<VM>>();
+const activeComponents = new WeakMap<ComponentConstructor, Set<VM>>();
+const activeStyles = new WeakMap<StylesheetFactory, Set<VM>>();
+
+function rehydrateHotTemplate(tpl: Template): boolean {
+    const list = activeTemplates.get(tpl);
+    list?.forEach((vm) => {
+        if (isFalse(vm.isDirty)) {
+            // forcing the vm to rehydrate in the next tick
+            markComponentAsDirty(vm);
+            scheduleRehydration(vm);
+        }
+    });
+    return true;
+}
+
+function rehydrateHotStyle(style: StylesheetFactory): boolean {
+    const list = activeStyles.get(style);
+    list?.forEach((vm) => {
+        if (isFalse(vm.isDirty)) {
+            // hacky way to force the styles to get recomputed
+            // by replacing the value of old template, which is used
+            // during the rendering process. If the template returned
+            // by render() is different from the previous stored template
+            // the styles will be reset, along with the content of the
+            // shadow, this way we can guarantee that the styles will be
+            // recalculated, and applied.
+            vm.cmpTemplate = () => [];
```

Comment:
Should this be moved outside the if block? If the vm was marked dirty before entering this routine, then its `cmpTemplate` will not be reset and it will continue to use the old `cmpTemplate`?

Reply:
good catch!